feat: Enable cluster wide cancel-in-progress setting via PAC ConfigMap

zakisk opened the pull request:
Cancel-in-progress now can be enabled via PAC configMap using `enable-cancel-in-progress-on-pull-requests` and `enable-cancel-in-progress-on-push` fields for pull request and push events respectively. if a PipelineRun is having `cancel-in-progress` annotation specified explicitly then it will override setting from PAC configMap.


# Changes <!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

<!-- Describe your changes here- ideally you can get that description straight from your descriptive commit message(s)! -->

# Submitter Checklist

- [x] 📝 Ensure your commit message is clear and informative. Refer to the How to write a git commit message guide. Include the commit message in the PR body rather than linking to an external site (e.g., Jira ticket).

- [x] ♽ Run make test lint before submitting a PR to avoid unnecessary CI processing. Consider installing [pre-commit](https://pre-commit.com/) and running pre-commit install in the repository root for an efficient workflow.

- [x] ✨ We use linters to maintain clean and consistent code. Run make lint before submitting a PR. Some linters offer a --fix mode, executable with make fix-linters (ensure [markdownlint](https://github.com/DavidAnson/markdownlint) and [golangci-lint](https://github.com/golangci/golangci-lint) are installed).

- [x] 📖 Document any user-facing features or changes in behavior.

- [x] 🧪 While 100% coverage isn't required, we encourage unit tests for code changes where possible.

- [x] 🎁 If feasible, add an end-to-end test. See [README](https://github.com/openshift-pipelines/pipelines-as-code/blob/main/test/README.md) for details.

- [ ] 🔎 Address any CI test flakiness before merging, or provide a valid reason to bypass it (e.g., token rate limitations).

- If adding a provider feature, fill in the following details:

  - [x] GitHub App
  - [x] GitHub Webhook
  - [x] Gitea/Forgejo
  - [x] GitLab
  - [x] Bitbucket Cloud
  - [ ] Bitbucket Data Center

  (update the provider documentation accordingly)

zakisk committed:
feat: Enable cluster wide setting for cancel-in-progress

currently, cancel-in-progress can be configured only on
PipelineRun level via cancel-in-progress annotation.
This PR introduces new setting in Pipelines as Code
ConfigMap with enable-cancel-in-progress-on-pull-requests and
enable-cancel-in-progress-on-push to control cancel-in-progress
functionality on deployment level.

when enable-cancel-in-progress-on-pull-requests is set to true
all the PipelineRun will be cancelled on PR close and on PR
update when a commit is pushed on the head branch of PR and
same for push setting. a PipelineRun will override configmap
setting if it has cancel-in-progress annotation explicitly
and PAC will do accordingly.

https://issues.redhat.com/browse/SRVKP-6910

Signed-off-by: Zaki Shaikh <zashaikh@redhat.com>

zakisk commented:
linteres is failing on 😕 
```
pkg/pipelineascode/cancel_pipelineruns.go:56: NotIn ==> noting, not in, notion
pkg/pipelineascode/cancel_pipelineruns_test.go:1525: NotIn ==> noting, not in, notion
pkg/pipelineascode/cancel_pipelineruns_test.go:1526: notin ==> noting, not in, notion
```

chmouel commented:
> linteres is failing on 😕
> 
> ```
> pkg/pipelineascode/cancel_pipelineruns.go:56: NotIn ==> noting, not in, notion
> pkg/pipelineascode/cancel_pipelineruns_test.go:1525: NotIn ==> noting, not in, notion
> pkg/pipelineascode/cancel_pipelineruns_test.go:1526: notin ==> noting, not in, notion
> ```

can you send a pr ?

zakisk commented:
> can you send a pr ?

yes

zakisk committed:
fix: Match source and head branch in full path

before this, when source or head branch in cancel-in-progress
matching code having full path weren't being matched.
e.g. refs/heads/branch

Signed-off-by: Zaki Shaikh <zashaikh@redhat.com>

zakisk commented:
@chmouel codespell PR #2055

chmouel commented on test/github_pullrequest_test.go:
```diff
@@ -478,6 +478,117 @@ func TestGithubPullRequestNoPipelineRunCancelledOnPRClosed(t *testing.T) {
+	ctx, runcnx, _, _, err := tgithub.Setup(ctx, false, false)
+	assert.NilError(t, err)
+
+	patchData := map[string]interface{}{
```
consider using https://github.com/openshift-pipelines/pipelines-as-code/blob/fbacd0eb4/test/pkg/configmap/configmap.go#L15

zakisk replied:
ah, thanks! that's exactly what I am doing.

chmouel commented on test/pkg/github/setup.go:
```diff
@@ -106,3 +112,58 @@ func Setup(ctx context.Context, onSecondController, viaDirectWebhook bool) (cont
 	return ctx, run, e2eoptions, gprovider, nil
 }
+
+func PatchPACConfigMap(ctx context.Context, run *params.Run, patch map[string]interface{}) (*corev1.ConfigMap, error) {
```
same as before

chmouel reviewed (commented):
this looks good to me, just a comment about some duplication..

I'd like to enable this by default on our dogfooding cluster, please ping me to let you know how to do that
